src/main/java/htsjdk/samtools/util/AsyncWriterPool.java

```diff
@@ -0,0 +1,253 @@
+        private final AtomicReference<Future<Void>> currentTask = new AtomicReference<>(null);
+
+        /**
+         * Exchange a class implementing {@link Writer} for a {@code PooledWriter}.
+         *
+         * @param writer         a class implementing {@link Writer}
+         * @param queue          a queue to use for this writer, bounded or unbounded.
+         * @param writeThreshold the minimum number of items needed before scheduling a thread for writing.
+         * @throws IllegalArgumentException if writeThreshold <= 0 or if writeThreshold > queue.remainingCapacity()
+         */
+        private PooledWriter(final Writer<A> writer, final BlockingQueue<A> queue, final int writeThreshold) {
+            if (writeThreshold <= 0) throw new IllegalArgumentException("writeThreshold must be >= 1");
+            if (writeThreshold > queue.remainingCapacity())
+                throw new IllegalArgumentException("writeThreshold can't be larger then queue capacity.");
+
+            this.writer = writer;
+            this.queue = queue;
+            this.writeThreshold = writeThreshold;
+        }
+
+        /**
+         * Non-blocking check of the previously completed task. This will check the future if it is not null and if it
+         * isDone. If the task generated any exception, these will be re-thrown.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void asyncCheckAndRethrow() {
+            if (this.currentTask.get() != null && this.currentTask.get().isDone()) {
+                this.checkAndRethrow();
+            }
+        }
+
+        /**
+         * Blocking check of the previously completed task. This will wait for the future to complete if it is not null.
+         * If the task generated any exception, these will be re-thrown.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void syncCheckAndRethrow() {
+            if (this.currentTask.get() != null) {
+                this.checkAndRethrow();
+            }
+        }
+
+        /**
+         * Collect any exceptions thrown by the task, close writer, set the current task to null, and rethrow the
+         * caught exceptions.
+         * <p>
+         * This method should not be called directly and should only be used through the {@code sync} and {@code async}
+         * variants that go by the same name.
+         *
+         * @throws RuntimeException if any exception was raised during the writing on the receiving thread
+         */
+        private void checkAndRethrow() {
+            try {
+                this.currentTask.get().get();
+            } catch (CancellationException | InterruptedException | ExecutionException e) {
+                this.isClosed.set(true);
+                throw new RuntimeException("Exception while writing records asynchronously", e);
```

Comment:
In the case of an ExecutionException we might want to unwrap it and rethrow the exception that caused the ExecutionException.

Reply:
Is there a way to do that without making `ExecutionException` a checked exception?